ciao-controller/internal/datastore/datastore.go

```diff
@@ -1406,6 +1406,29 @@ func (ds *Datastore) AddBlockDevice(device types.BlockData) error {
 	return nil
 }
 
+func (ds *Datastore) DeleteBlockDevice(ID string) error {
+	// lock both tenants and devices maps
+	var err error
+
+	ds.bdLock.Lock()
+	ds.tenantsLock.Lock()
+
+	dev, ok := ds.blockDevices[ID]
+	if ok {
+		delete(ds.blockDevices, ID)
+		delete(ds.tenants[dev.TenantID].devices, ID)
+
+		go ds.db.deleteBlockData(ID)
```

Comment:
it's ok to have this happen asynchronously. The reason it's ok is that each block device is only ever operated on if it exists in the cache. If it doesn't exist in the cache, it is treated as a failure. the sqlite db is really just used as long term persistent datastore. There *is* a possibility that if ciao-controller is terminated prior to the go routine which is doing the asynchronous update has completed, we'll leave behind a volume that was meant to be deleted that will then be repopulated into the cache with ciao-controller is restarted. but that would happen regardless of any additional measures I could take with wait groups or channels.